File: ts/src/bake.ts
/**
 * Bake: validate + resolve, then seal the values with the schema into a
 * Baked snapshot in canonical wire form. Mirrors the Go reference bake.go.
 */

import { create, equals } from "@bufbuild/protobuf";
import Mustache from "mustache";
import { fieldIsActive, refDefKey, selectVariant } from "./compute.js";
import type { Engine } from "./engine.js";
import type { ValidationResult } from "./gen/schemapb/errors_pb.js";
import type { Baked, Filled } from "./gen/schemapb/runtime_pb.js";
import { BakedSchema } from "./gen/schemapb/runtime_pb.js";
import type { Schema, Schema_Field } from "./gen/schemapb/schema_pb.js";
import { SchemaSchema } from "./gen/schemapb/schema_pb.js";
import type { StructValue, Value } from "./gen/schemapb/value_pb.js";
import { StructValueSchema } from "./gen/schemapb/value_pb.js";
import { displayString, type RenderContext, type RenderField, renderField } from "./render.js";
import type { TemplateName } from "./typed.js";
import { resultBlocking, validate } from "./validate.js";
import {
  canonicalStruct,
  canonicalValue,
  fromNative,
  isNativeStruct,
  type Native,
  type NativeStruct,
  structToNative,
} from "./value.js";

export interface BakeOutcome {
  baked?: Baked;
  result: ValidationResult;
}

/**
 * Validates and resolves values, then seals them (canonical wire variants).
 * On a blocking failure `baked` is absent; warnings do not block.
 */
export function bake(e: Engine, values: NativeStruct): BakeOutcome {
  const result = validate(e, values);
  if (resultBlocking(result)) {
    return { result };
  }
  const baked = create(BakedSchema, {
    schema: e.schema,
    values: canonicalEngineStruct(e, values),
  });
  return { baked, result };
}

/** Projects a resolved native form into canonical wire variants. */
function canonicalEngineStruct(e: Engine, values: NativeStruct): StructValue {
  const fields: Record<string, Value> = {};
  for (const [name, val] of Object.entries(values)) {
    const f = e.schema.fields.find((x) => x.name === name);
    fields[name] = canonicalTop(e, f, val);
  }
  return create(StructValueSchema, { fields });
}

function canonicalTop(e: Engine, f: Schema_Field | undefined, val: Native): Value {
  if (f === undefined) {
    return fromNative(val);
  }
  if (f.kind.case === "ref") {
    const def = e.schema.defs[refDefKey(f.kind.value)];
    if (def !== undefined && isNativeStruct(val)) {
      return canonicalStruct(def, val);
    }
    return fromNative(val);
  }
  if (f.kind.case === "oneOf") {
    const sel = selectVariant(f.kind.value, val);
    if (sel !== undefined) {
      return canonicalStruct(sel[0], sel[1]);
    }
    return fromNative(val);
  }
  try {
    return canonicalValue(f, val);
  } catch {
    return fromNative(val);
  }
}

/**
 * Layers overrides onto a baked form (objects merge recursively, lists
 * append unless replaceLists, scalars overwrite) and re-seals on this
 * engine.
 */
export function merge(
  e: Engine,
  baked: Baked,
  overrides: StructValue,
  replaceLists: boolean,
): BakeOutcome {
  const base = structToNative(baked.values);
  const over = structToNative(overrides);
  return bake(e, mergeStructs(base, over, replaceLists));
}

function mergeStructs(dst: NativeStruct, src: NativeStruct, replaceLists: boolean): NativeStruct {
  const out: NativeStruct = { ...dst };
  for (const [k, sv] of Object.entries(src)) {
    const dv = out[k];
    if (dv !== undefined) {
      if (isNativeStruct(dv) && isNativeStruct(sv)) {
        out[k] = mergeStructs(dv, sv, replaceLists);
        continue;
      }
      if (!replaceLists && Array.isArray(dv) && Array.isArray(sv)) {
        out[k] = [...dv, ...sv];
        continue;
      }
    }
    out[k] = sv;
  }
  return out;
}

/** Whether the baked schema is identical in content to s. */
export function bakedMatches(baked: Baked, s: Schema): boolean {
  return baked.schema !== undefined && equals(SchemaSchema, baked.schema, s);
}

/**
 * Validates and seals an inline Filled. Throws when the Filled references a
 * schema by id (a registry must resolve it first).
 */
export function filledBake(filled: Filled, compile: (s: Schema) => Engine): BakeOutcome {
  const source = filled.schema?.source;
  if (source?.case !== "schema") {
    throw new Error(
      "schemapb: Filled bake requires an inline schema (id refs resolve via a registry)",
    );
  }
  const e = compile(source.value);
  return bake(e, structToNative(filled.values));
}

/**
 * Renders a schema-carried Mustache template against resolved values.
 * Returns undefined when the template does not exist.
 */
export function render(e: Engine, name: TemplateName, values: NativeStruct): string | undefined {
  const tmpl = e.schema.templates[name as string];
  if (tmpl === undefined) {
    return undefined;
  }
  return Mustache.render(tmpl, buildRenderContext(e, values));
}

/**
 * Builds the contract render context (fields / groups / values with
 * precomputed display forms); inactive fields are excluded entirely.
 */
export function buildRenderContext(e: Engine, values: NativeStruct): RenderContext {
  const fields: RenderField[] = [];
  const groups: { name: string; fields: RenderField[] }[] = [];
  const groupIdx = new Map<string, number>();

  for (const f of e.schema.fields) {
    if ((f.when ?? "") !== "" && !fieldIsActive(e, f, values, f.name, undefined)) {
      continue;
    }
    const rf = renderField(f, values);
    fields.push(rf);
    const g = f.group ?? "";
    let i = groupIdx.get(g);
    if (i === undefined) {
      i = groups.length;
      groupIdx.set(g, i);
      groups.push({ name: g, fields: [] });
    }
    groups[i]?.fields.push(rf);
  }

  const display: Record<string, string> = {};
  for (const [name, val] of Object.entries(values)) {
    display[name] = val === null ? "" : displayString(val);
  }
  return { fields, groups, values: display };
}

/** Renders a Baked snapshot with a template of its embedded schema. */
export function renderBaked(e: Engine, baked: Baked, name: TemplateName): string | undefined {
  return render(e, name, structToNative(baked.values));
}

File: rust/src/bake.rs
//! Bake / merge / render, mirroring the Go reference bake.go + render.go.

use crate::compute::{field_is_active, ref_def_key, select_variant};
use crate::engine::Engine;
use crate::gen::schemapb::schema::field::Kind as K;
use crate::gen::schemapb::{Baked, Schema, StructValue, ValidationResult, Value};
use crate::render::{display_string, render_field, RenderContext, RenderField, RenderGroup};
use crate::validate::{result_blocking, validate};
use crate::value::{
    canonical_struct, canonical_value, from_native, struct_to_native, Native, NativeStruct,
    SchemaField,
};

pub struct BakeOutcome {
    pub result: ValidationResult,
    pub baked: Option<Baked>,
}

/// Validate + resolve, then seal in canonical wire form.
pub(crate) fn bake(e: &Engine, values: &mut NativeStruct) -> BakeOutcome {
    let result = validate(e, values);
    if result_blocking(&result) {
        return BakeOutcome {
            result,
            baked: None,
        };
    }
    let baked = Baked {
        schema: Some(e.schema.clone()),
        values: Some(canonical_engine_struct(e, values)),
    };
    BakeOutcome {
        result,
        baked: Some(baked),
    }
}

fn canonical_engine_struct(e: &Engine, values: &NativeStruct) -> StructValue {
    let fields = values
        .iter()
        .map(|(name, val)| {
            let f = e.schema.fields.iter().find(|x| &x.name == name);
            (name.clone(), canonical_top(e, f, val))
        })
        .collect();
    StructValue { fields }
}

fn canonical_top(e: &Engine, f: Option<&SchemaField>, val: &Native) -> Value {
    let Some(f) = f else {
        return from_native(val);
    };
    match f.kind.as_ref() {
        Some(K::Ref(r)) => {
            if let (Some(def), Some(m)) = (e.schema.defs.get(&ref_def_key(r)), val.as_struct()) {
                if let Ok(v) = canonical_struct(def, m) {
                    return v;
                }
            }
            from_native(val)
        }
        Some(K::OneOf(oo)) => {
            if let (Some(variant), Some(m)) = (select_variant(oo, val), val.as_struct()) {
                if let Ok(v) = canonical_struct(variant, m) {
                    return v;
                }
            }
            from_native(val)
        }
        _ => canonical_value(f, val).unwrap_or_else(|_| from_native(val)),
    }
}

/// Layers overrides onto a baked form and re-seals on this engine.
#[must_use]
pub(crate) fn merge(
    e: &Engine,
    baked: &Baked,
    overrides: &StructValue,
    replace_lists: bool,
) -> BakeOutcome {
    let base = struct_to_native(baked.values.as_ref());
    let over = struct_to_native(Some(overrides));
    let mut merged = merge_structs(&base, &over, replace_lists);
    bake(e, &mut merged)
}

fn merge_structs(dst: &NativeStruct, src: &NativeStruct, replace_lists: bool) -> NativeStruct {
    let mut out = dst.clone();
    for (k, sv) in src {
        match (out.get(k), sv) {
            (Some(Native::Struct(dm)), Native::Struct(sm)) => {
                out.insert(
                    k.clone(),
                    Native::Struct(merge_structs(dm, sm, replace_lists)),
                );
            }
            (Some(Native::List(dl)), Native::List(sl)) if !replace_lists => {
                let mut joined = dl.clone();
                joined.extend(sl.iter().cloned());
                out.insert(k.clone(), Native::List(joined));
            }
            _ => {
                out.insert(k.clone(), sv.clone());
            }
        }
    }
    out
}

/// Whether the baked schema is identical in content to `s`.
#[must_use]
pub(crate) fn baked_matches(baked: &Baked, s: &Schema) -> bool {
    baked.schema.as_ref() == Some(s)
}

/// Renders a schema-carried Mustache template against resolved values.
#[must_use]
pub(crate) fn render(e: &Engine, name: &str, values: &NativeStruct) -> Option<String> {
    e.render_template(name, &build_render_context(e, values))
}

/// Renders a Baked snapshot with a template of its embedded schema.
#[must_use]
pub(crate) fn render_baked(e: &Engine, baked: &Baked, name: &str) -> Option<String> {
    render(e, name, &struct_to_native(baked.values.as_ref()))
}

/// The contract render context; inactive fields excluded entirely.
#[must_use]
pub(crate) fn build_render_context(e: &Engine, values: &NativeStruct) -> RenderContext {
    let mut fields: Vec<RenderField> = Vec::new();
    let mut groups: Vec<RenderGroup> = Vec::new();
    let mut group_idx: std::collections::HashMap<String, usize> = std::collections::HashMap::new();

    for f in &e.schema.fields {
        if !f.when.as_deref().unwrap_or("").is_empty()
            && !field_is_active(e, f, values, &f.name, None)
        {
            continue;
        }
        let rf = render_field(f, values);
        fields.push(rf.clone());
        let g = f.group.clone().unwrap_or_default();
        let i = *group_idx.entry(g.clone()).or_insert_with(|| {
            groups.push(RenderGroup {
                name: g,
                fields: Vec::new(),
            });
            groups.len() - 1
        });
        groups[i].fields.push(rf);
    }

    let display = values
        .iter()
        .map(|(name, val)| {
            (
                name.clone(),
                if val.is_null() {
                    String::new()
                } else {
                    display_string(val)
                },
            )
        })
        .collect();
    RenderContext {
        fields,
        groups,
        values: display,
    }
}

impl Baked {
    /// Whether the embedded schema is identical in content to `s`.
    #[must_use]
    pub fn matches(&self, s: &Schema) -> bool {
        baked_matches(self, s)
    }
}
